client/consensus/manual-seal/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+};
+use sp_blockchain::HeaderBackend;
+use client_api::backend::Backend as ClientBackend;
+use futures::prelude::*;
+use hash_db::Hasher;
+use transaction_pool::{BasicPool, txpool};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+pub mod rpc;
+mod error;
+
+pub use error::Error;
+pub use rpc::{EngineCommand, CreatedBlock};
+
+/// The synchronous block-import worker of the engine.
+pub struct ManualSealBlockImport<I> {
+	inner: I,
+}
+
+impl<I> From<I> for ManualSealBlockImport<I> {
+	fn from(i: I) -> Self {
+		ManualSealBlockImport { inner: i }
+	}
+}
+
+impl<B: BlockT, I: BlockImport<B>> BlockImport<B> for ManualSealBlockImport<I> {
+	type Error = I::Error;
+
+	fn check_block(&mut self, block: BlockCheckParams<B>) -> Result<ImportResult, Self::Error>
+	{
+		self.inner.check_block(block)
+	}
+
+	fn import_block(
+		&mut self,
+		block: BlockImportParams<B>,
+		cache: HashMap<CacheKeyId, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		// TODO: strip out post-digest.
+
+		self.inner.import_block(block, cache)
+	}
+}
+
+/// The verifier for the manual seal engine; instantly finalizes.
+struct ManualSealVerifier;
+
+impl<B: BlockT> Verifier<B> for ManualSealVerifier {
+	fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justification: Option<Justification>,
+		body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let import_params = BlockImportParams {
+			origin,
+			header,
+			justification,
+			post_digests: Vec::new(),
+			body,
+			finalized: true,
+			auxiliary: Vec::new(),
+			fork_choice: ForkChoiceStrategy::LongestChain,
+			allow_missing_state: false,
+			import_existing: false,
+		};
+
+		Ok((import_params, None))
+	}
+}
+
+/// Instantiate the import queue for the manual seal consensus engine.
+pub fn import_queue<B: BlockT>(block_import: BoxBlockImport<B>) -> BasicQueue<B>
+{
+	BasicQueue::new(
+		ManualSealVerifier,
+		block_import,
+		None,
+		None,
+	)
+}
+
+/// Creates the background authorship task for the manual seal engine.
+pub async fn run_manual_seal<B, CB, E, A, C, S, H>(
+	mut block_import: BoxBlockImport<B>,
+	mut env: E,
+	back_end: Arc<CB>,
+	basic_pool: Arc<BasicPool<A, B>>,
+	mut seal_block_channel: S,
+	select_chain: C,
+	inherent_data_providers: inherents::InherentDataProviders,
+)
+	where
+		B: BlockT + 'static,
+		H: Hasher<Out=<B as BlockT>::Hash>,
+		CB: ClientBackend<B, H> + 'static,
+		E: Environment<B> + 'static,
+		E::Error: std::fmt::Display,
+		<E::Proposer as Proposer<B>>::Error: std::fmt::Display,
+		A: txpool::ChainApi<Block=B, Hash=<B as BlockT>::Hash> + 'static,
+		S: Stream<Item=EngineCommand<<B as BlockT>::Hash>> + Unpin + 'static,
+		C: SelectChain<B> + 'static,
+{
+	let pool = basic_pool.pool();
+
+	while let Some(command) = seal_block_channel.next().await {
+		match command {
+			EngineCommand::SealNewBlock {
+				create_empty,
+				parent_hash,
+				sender
+			} => {
```

Comment:
refactor: wouldn't it be better to have a separate function for each of these blocks?

Reply:
makes sense.